File: mobile/android/src/main/java/com/rpm/clynx/utility/TimeAgoUtils.java
package com.rpm.clynx.utility;

import android.os.Build;
import androidx.annotation.RequiresApi;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

public class TimeAgoUtils {
    @RequiresApi(api = Build.VERSION_CODES.O)
    public static String calculateTimeAgo(String localDateString) {
        try {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MMM-dd,h:mm a", Locale.getDefault());
            Date localDate = sdf.parse(localDateString);

            // Get the current local time
            long currentTimeMillis = System.currentTimeMillis();
            Date currentDate = new Date(currentTimeMillis);

            Calendar calendar1 = Calendar.getInstance();
            calendar1.setTime(localDate);

            Calendar calendar2 = Calendar.getInstance();
            calendar2.setTime(currentDate);

            // Calculate the time difference in milliseconds
            long timeDifferenceMillis = calendar2.getTimeInMillis() - calendar1.getTimeInMillis();
            long secondsDifference = TimeUnit.MILLISECONDS.toSeconds(timeDifferenceMillis);
            long minutesDifference = TimeUnit.MILLISECONDS.toMinutes(timeDifferenceMillis);
            long hoursDifference = TimeUnit.MILLISECONDS.toHours(timeDifferenceMillis);

            if (secondsDifference < 60) {
                if (secondsDifference <= 5) {
                    return "Just now";
                } else {
                    return secondsDifference + " seconds ago";
                }
            } else if (minutesDifference < 60) {
                // Round the minutes to the nearest unit
                minutesDifference = Math.round((float) timeDifferenceMillis / TimeUnit.MINUTES.toMillis(1));
                return minutesDifference + " minutes ago";
            } else if (hoursDifference < 24) {
                // Round the hours to the nearest unit
                hoursDifference = Math.round((float) timeDifferenceMillis / TimeUnit.HOURS.toMillis(1));
                return hoursDifference + " hours ago";
            }

            // Calculate days difference, considering the time
            long daysDifference = TimeUnit.MILLISECONDS.toDays(timeDifferenceMillis);

            if (daysDifference == 0 && hoursDifference > 24) {
                daysDifference = 1;
            }

            if (daysDifference < 7) {
                return calendar1.getDisplayName(Calendar.DAY_OF_WEEK, Calendar.LONG, Locale.getDefault());
            } else if (daysDifference >= 7 && daysDifference <= 14) {
                return "1 week ago";
            } else if (daysDifference > 14 && daysDifference <= 21) {
                return "2 weeks ago";
            } else if (daysDifference > 21 && daysDifference <= 28) {
                return "3 weeks ago";
            } else {
                int yearDiff = calendar2.get(Calendar.YEAR) - calendar1.get(Calendar.YEAR);
                int monthDiff = calendar2.get(Calendar.MONTH) - calendar1.get(Calendar.MONTH);

                if (yearDiff > 0 || monthDiff > 1) {
                    return monthDiff + " months ago";
                } else if (yearDiff == 0 && monthDiff == 1) {
                    return "1 month ago";
                }
            }

            // Return a default value if necessary
            return "Some default text";
        } catch (ParseException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: mobile/ios/rpm_swift/RPM/View/Common/TimeAgoUtils.swift
//
//  TimeAgoUtils.swift
//  RPM
//
//  Created by Tesplabs on 08/11/23.
//

import SwiftUI

struct TimeAgoUtils {
    static func calculateTimeAgo(localDateString: String) -> String {
        
        // First, create a DateFormatter for parsing the input string
        let inputDateFormatter = DateFormatter()
        inputDateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        inputDateFormatter.timeZone = TimeZone(identifier: "UTC")
        
        guard let utcDate = inputDateFormatter.date(from: localDateString) else {
            print("Failed to parse the input date string.")
            return "Invalid Date"
        }
        
        // Convert UTC date to local time zone
        let localDate = utcDate.addingTimeInterval(TimeInterval(TimeZone.current.secondsFromGMT(for: utcDate)))
        
        print("localDate: \(localDate)")
      
        
        // Set the calendar and time zone to the system's local time
        let calendar = Calendar.current
        
        // Ensure the current date is obtained in the local time zone
        let currentDate = Date().addingTimeInterval(TimeInterval(TimeZone.current.secondsFromGMT()))
        
        let dateFormatterNW = DateFormatter()
        dateFormatterNW.dateFormat = "yyyy-MM-dd HH:mm:ss"
        dateFormatterNW.timeZone = TimeZone.current  // Set the desired time zone
        let formattedDate = dateFormatterNW.string(from: currentDate)
        
        print("Current Date: \(currentDate)")
        print("formattedCDate: \(formattedDate)")
        
        // Calculate the time difference in seconds
     
        let timeDifference = calendar.dateComponents([.second, .minute, .hour, .day], from: localDate, to: currentDate)
        print("Time Difference: \(timeDifference)")
        
        
        let secondsDifference = timeDifference.second ?? 0
        let minutesDifference = timeDifference.minute ?? 0
        let hoursDifference = timeDifference.hour ?? 0
        let daysDifference = timeDifference.day ?? 0
        
        let weeksDifference = timeDifference.weekOfYear ?? 0
              let monthsDifference = timeDifference.month ?? 0
        
        print("secondsDifference: \(secondsDifference)")
        print("minutesDifference: \(minutesDifference)")
        print("hoursDifference: \(hoursDifference)")
        print("daysDifference: \(daysDifference)")
        
        // Prioritize larger time units
        
        if monthsDifference > 0 {
                  return "\(monthsDifference) \(monthsDifference == 1 ? "month" : "months") ago"
              } else if weeksDifference > 0 {
                  return "\(weeksDifference) \(weeksDifference == 1 ? "week" : "weeks") ago"
              }
        
        
        
        else if daysDifference == 0 && hoursDifference > 0 && hoursDifference < 24{
            
            let roundedHours: Int
               if minutesDifference > 30 {
                   // If the minute difference is greater than 30, round up the hours
                   roundedHours = hoursDifference + 1
               } else {
                   roundedHours = hoursDifference
               }
               
               return "\(roundedHours) \(roundedHours == 1 ? "hour" : "hours") ago"
            
            
            
             
           } else if daysDifference == 0 && hoursDifference == 0 && minutesDifference > 0 && minutesDifference < 60{
               return "\(minutesDifference) \(minutesDifference == 1 ? "minute" : "minutes") ago"
           }
           
           else if daysDifference == 0 && hoursDifference == 0 && minutesDifference == 0 && secondsDifference > 0 && secondsDifference < 60 {
               return "\(secondsDifference) \(secondsDifference == 1 ? "second" : "seconds") ago"
           }
           
        
        else if daysDifference < 7 && daysDifference >= 1 {
            let localDayOfWeek = calendar.component(.weekday, from: localDate)
            let correctedLocalDayOfWeek = (localDayOfWeek + 5) % 7 // Adjust to start from Monday
            let dayOfWeekSymbol = calendar.weekdaySymbols[correctedLocalDayOfWeek]
            print("localDayOfWeek: \(localDayOfWeek), correctedLocalDayOfWeek: \(correctedLocalDayOfWeek), dayOfWeekSymbol: \(dayOfWeekSymbol)")
            return dayOfWeekSymbol
        }


        else if daysDifference  >= 7 && daysDifference <= 14 {
            return "1 week ago"
        } else if daysDifference  >= 14 && daysDifference  <= 21 {
            return "2 weeks ago"
        } else if daysDifference  >= 21 && daysDifference <= 28 {
            return "3 weeks ago"
        }
  
              else {
                    let yearDiff = calendar.component(.year, from: currentDate) - calendar.component(.year, from: localDate)
                    let monthDiff = calendar.component(.month, from: currentDate) - calendar.component(.month, from: localDate)

                    if yearDiff > 0 || monthDiff > 1 {
                        return "\(monthDiff) months ago"
                    } else if yearDiff == 0 && monthDiff == 1 {
                        return "1 month ago"
                    }
                }
            
     
        return "Some default text"
    }
}

